src/components/app.rs

```diff
@@ -8,47 +9,127 @@ use super::{
     users_cache::{self, UsersCacheComponent},
 };
 
+use std::sync::Mutex;
+
 pub struct AppComponents {
     pub health: HealthComponent,
     pub synapse: SynapseComponent,
     pub config: Config,
     pub db: DatabaseComponent,
-    pub user_cache: UsersCacheComponent,
+    pub users_cache: Mutex<UsersCacheComponent>,
 }
 
-impl AppComponents {
-    pub async fn new(custom_config: Option<Config>) -> Self {
-        if let Err(_) = env_logger::try_init() {
-            log::debug!("Logger already init")
-        }
+pub struct CustomComponents {
+    pub synapse: Option<SynapseComponent>,
+    pub db: Option<DatabaseComponent>,
+    pub users_cache: Option<UsersCacheComponent>,
+    pub redis: Option<Redis>,
+}
 
-        // Initialize components
+impl AppComponents {
+    pub async fn new(
+        custom_config: Option<Config>,
+        custom_components: Option<CustomComponents>,
+    ) -> Self {
         let config = custom_config
             .unwrap_or_else(|| Config::new().expect("Couldn't read the configuration"));
+        if custom_components.is_none() {
+            AppComponents::default(config).await
+        } else {
+            // For testing propouses mainly
+            let custom = custom_components.unwrap();
+            AppComponents::custom(config, custom).await
+        }
+    }
 
-        let mut health = HealthComponent::default();
-        let synapse = SynapseComponent::new(config.synapse.url.clone());
-        let mut db = DatabaseComponent::new(&config.db);
-
-        if let Err(err) = db.run().await {
-            log::debug!("Error on running the DB: {:?}", err);
-            panic!("Unable to run the DB")
+    async fn default(config: Config) -> Self {
+        if env_logger::try_init().is_err() {
+            log::debug!("Logger already init")
         }
 
+        // Initialize components
+        let synapse = AppComponents::init_synapse_component(config.synapse.url.clone());
+        let db = AppComponents::init_db_component(&config.db).await;
         let redis = Redis::new_and_run(&config.redis);
 
         // TODO: Should we refactor HealthComponent to avoid cloning structs?
-        health.register_component(Box::new(db.clone()), "database".to_string());
-        health.register_component(Box::new(redis.clone()), "redis".to_string());
+        let health = AppComponents::init_health_component(db.clone(), redis.clone());
+
         let users_cache_instance =
-            users_cache::UsersCacheComponent::new(redis, config.cache_hashing_key.clone());
+            AppComponents::init_users_cache(redis, config.cache_hashing_key.clone());
```

Comment:
nit: Could we replace the type name here as follows?
```suggestion
            Self::init_users_cache(redis, config.cache_hashing_key.clone());
```